### PR TITLE
Yamux implementation

### PR DESCRIPTION
**Pull request by ianopolous**
Part of https://github.com/libp2p/jvm-libp2p/issues/272

The spec is here https://github.com/hashicorp/yamux/blob/master/spec.md

### CODE REVIEW

File: libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            ctx.write(YamuxFrame(msg.id, YamuxType.WINDOW_UPDATE, 0, delta))
+            ctx.flush()
+        }
+        childRead(msg.id, msg.data!!)
+    }
+
+    fun handleWindowUpdate(msg: YamuxFrame) {
+        handleFlags(msg)
+        val size = msg.lenData
+        sendWindow.addAndGet(size)
```
Comment (Nashatyrev) [inline libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt:95]: Isn't window tracked on per-stream basis?

**Review by Nashatyrev — CHANGES_REQUESTED** [review]
Some points to fix and some potential points to discuss: 
- Track window per stream?
- How to handle sending when window size is 0?

File: libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxFrameCodec.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+     * @param out the list to write the extracted frame to.
+     */
+    override fun decode(ctx: ChannelHandlerContext, msg: ByteBuf, out: MutableList<Any>) {
+        while (msg.isReadable) {
+            val readerIndex = msg.readerIndex()
+            msg.readByte(); // version always 0
+            val type = msg.readUnsignedByte()
+            val flags = msg.readUnsignedShort()
+            val streamId = msg.readInt()
+            val lenData = msg.readInt()
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxFrameCodec.kt]: Done

File: libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxFrameCodec.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            if (type.toInt() != YamuxType.DATA) {
+                val yamuxFrame = YamuxFrame(MuxId(ctx.channel().id(), streamId.toLong(), isInitiator.xor(streamId % 2 == 1).not()), type.toInt(), flags, lenData)
+                out.add(yamuxFrame)
+                return
+            }
+            if (lenData < 0) {
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxFrameCodec.kt]: Removed

File: libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt (hunk 1 above)
Comment (ianopolous) [reply libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt:101]: Yep, I've fixed that

File: libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            YamuxType.WINDOW_UPDATE -> handleWindowUpdate(msg)
+            YamuxType.PING -> handlePing(msg)
+            YamuxType.GO_AWAY -> onRemoteClose(msg.id)
+        }
+    }
+
+    fun handlePing(msg: YamuxFrame) {
+        val ctx = getChannelHandlerContext()
+        when (msg.flags) {
+            YamuxFlags.SYN -> ctx.write(YamuxFrame(MuxId(msg.id.parentId, 0, msg.id.initiator), YamuxType.PING, YamuxFlags.ACK, msg.lenData))
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt]: Yep I've fixed that

File: libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            YamuxFlags.ACK -> {}
+        }
+    }
+
+    fun handleFlags(msg: YamuxFrame) {
+        val ctx = getChannelHandlerContext()
+        if (msg.flags == YamuxFlags.SYN) {
+            // ACK the new stream
+            onRemoteOpen(msg.id)
+            ctx.write(YamuxFrame(msg.id, YamuxType.WINDOW_UPDATE, YamuxFlags.ACK, 0))
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxHandler.kt]: Fixed

File: libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxFrameCodec.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            val readerIndex = msg.readerIndex()
+            msg.readByte(); // version always 0
+            val type = msg.readUnsignedByte()
+            val flags = msg.readUnsignedShort()
+            val streamId = msg.readInt()
+            val lenData = msg.readInt()
+            if (type.toInt() != YamuxType.DATA) {
+                val yamuxFrame = YamuxFrame(MuxId(ctx.channel().id(), streamId.toLong(), isInitiator.xor(streamId % 2 == 1).not()), type.toInt(), flags, lenData)
+                out.add(yamuxFrame)
+                return
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply libp2p/src/main/kotlin/io/libp2p/mux/yamux/YamuxFrameCodec.kt]: Good catch! That's fixed now.

**Comment by ianopolous** [comment]
Thanks for the review @Nashatyrev ! I think that's all fixed now except the locking on empty window size. I'm open to suggestions for that.

**Comment by ianopolous** [comment]
I could just put the pending writes in a queue I guess and process them on the window update thread..

**Comment by ianopolous** [comment]
That would solve the threading, but then it's not propagating back the back pressure up the stack. Any ideas?

**Comment by Nashatyrev** [comment]
I'm not 100% sure how it is supposed to correctly handle backpressure in Netty.

My understanding: 
- from the application side backpressure is reported via `Channel.isWritable` attribute and `ChannelInboundHandler.channelWritabilityChanged` notification event. The application protocol should respect (in theory) that attribute and suspend writing when needed (`isWritable == false`). 
- from the Netty side (or may be OS TCP protocol implementation side or may be both) there should be a buffer (likely size bounded) which tolerate backpressure ignorance until some level. Probably when that buffer overflows the connection is abruptly closed.
- When the buffer overflows the connection is terminated 

So what we most likely should do is to switch `isWritable` on the _stream_ `Channel` when `sendWindow` switches to zero and back

With respect to buffering I see the following possible options: 
- maintain and limit buffer per stream to tolerate backpressure 
- maintain per-stream buffers but limit the total size per connection
- on buffer overflow abruptly close a stream or connection

**Comment by ianopolous** [comment]
Forgive my netty ignorance, but how do we set isWritable? I can only see a method channel.fireChannelWritabilityChanged()

In terms of DOS protection we should probably do the latter:
>     * maintain per-stream buffers but limit the total size per connection
>       
>       * on buffer overflow abruptly close connection

**Comment by Nashatyrev** [comment]
> Forgive my netty ignorance, but how do we set isWritable? I can only see a method channel.fireChannelWritabilityChanged()

Good question... I don't know 😃  

My intuition is that we may need to override the following methods in the `AbstractChildChannel`: 
- `boolean isWritable();`
- `long bytesBeforeUnwritable();`
- `long bytesBeforeWritable();`

Or we need to check out what's going on with the `ChannelOutboundBuffer` in child channels...

That looks like a topic which needs a deep dive. Thus I would suggest to just skip `isWritable` attribute support for this PR and file an issue for later investigation. 

So we should probably just support buffering with some configurable max size.

> In terms of DOS protection we should probably do the latter:

Agree 👍

**Comment by ianopolous** [comment]
There is one other thing I haven't implemented which is not in the spec, but is kind of a libp2p consensus - limiting unACKed streams to 256. But that can also be a future PR I think.

**Comment by ianopolous** [comment]
@Nashatyrev Could you clarify the remaining buffering stuff you'd like done?

**Comment by ianopolous** [comment]
> > @Nashatyrev Could you clarify the remaining buffering stuff you'd like done?
> 
> I believe the option below should work pretty fine:
> 
> > * maintain per-stream buffers but limit the total size per connection
> >   - on buffer overflow abruptly close connection

I've had a go at that. Let me know what you think.

**Comment by Nashatyrev** [comment]
Did some old code refactoring here. Could you please have a look: https://github.com/Peergos/jvm-libp2p/pull/7

**Review by Nashatyrev — APPROVED** [review]
Ok, let's give this PR a go, to not block further work. 
Meanwhile I'm planning to add some more generic Mux tests and tests specific to Yamux
Relates to #27

**Comment by ianopolous** [comment]
@Nashatyrev That sounds great. Some bugs we've found in real world use have suggested some good tests:
1) Open a connection and then have the recipient open a stream and ping down it. 
2) Do several thousand pings to test exceeding the initial window size.

**Comment by Nashatyrev** [comment]
@ianopolous I'm also not sure how `Go Away` yamux type should work 🤔 
From my perspective it should close the connection (not a stream)? If so then I don't understand its purpose - why not just close the connection?

**Comment by ianopolous** [comment]
My only guess is that it's used to communicate the type of error? To maybe not try reconnecting?

**Comment by Nashatyrev** [comment]
> My only guess is that it's used to communicate the type of error? To maybe not try reconnecting?

Yeah, probably. That makes sense. 
In eth2 we have RPC `Goodbye` call for that, but it's just outside of Libp2p domain

Then I guess we should remove that line 
```kotlin
            YamuxType.GO_AWAY -> onRemoteClose(msg.id)
```
cause it has no sense since `streamId == 0` always here

I think there should be some another handling mechanism if any other application needs to utilize that disconnect option. 
May be kind of API: 
```kotlin
Yamux.goAway(errorCode: Int)
Yamux.getRemoteGoAwayCode(): Int
```

**Pull request merged**